source-redshift-batch/driver.go

```diff
@@ -0,0 +1,791 @@
+		}
+	}
+
+	var keys []*discoveredPrimaryKey
+	for _, key := range keysByTable {
+		keys = append(keys, key)
+	}
+	return keys, nil
+}
+
+var databaseTypeToJSON = map[string]*jsonschema.Schema{
+	"integer":  {Type: "integer"},
+	"bigint":   {Type: "integer"},
+	"smallint": {Type: "integer"},
+	"boolean":  {Type: "boolean"},
+}
+
+var catalogNameSanitizerRe = regexp.MustCompile(`(?i)[^a-z0-9\-_.]`)
+
+func recommendedCatalogName(schema, table string) string {
+	var catalogName string
+	// Omit 'default schema' names for Postgres and SQL Server. There is
+	// no default schema for MySQL databases.
+	if schema == "public" || schema == "dbo" {
+		catalogName = table
+	} else {
+		catalogName = schema + "_" + table
+	}
+	return catalogNameSanitizerRe.ReplaceAllString(strings.ToLower(catalogName), "_")
+}
+
+// Validate checks that the configuration appears correct and that we can connect
+// to the database and execute queries.
+func (drv *BatchSQLDriver) Validate(ctx context.Context, req *pc.Request_Validate) (*pc.Response_Validated, error) {
+	// Perform discovery, which inherently validates that the config is well-formed
+	// and that we can connect to the database and execute (some) queries.
+	if _, err := drv.Discover(ctx, &pc.Request_Discover{
+		ConnectorType: req.ConnectorType,
+		ConfigJson:    req.ConfigJson,
+	}); err != nil {
+		return nil, err
+	}
+
+	// Unmarshal and validate resource bindings to make sure they're well-formed too.
+	var out []*pc.Response_Validated_Binding
+	for _, binding := range req.Bindings {
+		var res Resource
+		if err := pf.UnmarshalStrict(binding.ResourceConfigJson, &res); err != nil {
+			return nil, fmt.Errorf("parsing resource config: %w", err)
+		}
+
+		out = append(out, &pc.Response_Validated_Binding{
+			ResourcePath: []string{res.Name},
+		})
+	}
+	return &pc.Response_Validated{Bindings: out}, nil
+}
+
+// Pull is the heart of a capture connector and outputs a neverending stream of documents.
+func (drv *BatchSQLDriver) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) error {
+	var cfg Config
+	if err := pf.UnmarshalStrict(open.Capture.ConfigJson, &cfg); err != nil {
+		return fmt.Errorf("parsing endpoint config: %w", err)
+	}
+	cfg.SetDefaults()
+
+	var db, err = drv.Connect(stream.Context(), &cfg)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	var bindings []bindingInfo
+	for idx, binding := range open.Capture.Bindings {
+		var res Resource
+		if err := pf.UnmarshalStrict(binding.ResourceConfigJson, &res); err != nil {
+			return fmt.Errorf("parsing resource config: %w", err)
+		}
+		bindings = append(bindings, bindingInfo{
+			resource: res,
+			index:    idx,
+			stateKey: boilerplate.StateKey(binding.StateKey),
+		})
+	}
+
+	var state captureState
+	if open.StateJson != nil {
+		if err := pf.UnmarshalStrict(open.StateJson, &state); err != nil {
+			return fmt.Errorf("parsing state checkpoint: %w", err)
+		}
+	}
+
+	migrated, err := migrateState(&state, open.Capture.Bindings)
```

Comment:
We shouldn't need the migration stuff for a brand new capture. Doesn't specifically hurt anything to leave it in though if that makes the copy/paste easier, and we plan to refactor the commonality of these captures soon-ish.

Reply:
Yeah, the entire file driver.go is currently identical between Redshift and Postgres, and I'd prefer to keep it that way if possible to make the refactor just that much simpler.